RO syncs

icristescu opened the pull request:
A straightforward implementation for RO sync. I'll add more tests (concurrent ones too) so for now its a draft. 

The way I think `ro_sync` should be used is to catch exceptions for RO operations, and retry `ro_sync` in case of failure.  

Depends on https://github.com/mirage/index/pull/175

samoht commented on src/irmin-pack/dict.ml:
```diff
@@ -78,11 +80,12 @@ module Make (IO : IO.S) : S = struct
     let log_offset = IO.force_offset t.io in
     if log_offset > former_log_offset then refill ~from:former_log_offset t
 
+  let ro_sync t = if IO.readonly t.io then sync_offset t
```
out if curiosity, does it matter if we try to sync a RW instance? Why not allowing this?

pascutto replied:
The memory and disk versions of the `log` and `log_async` in a RW instance are always in sync (modulo flushing), so there is nothing to do in that case.

samoht replied:
so can we just call that function `sync`? (and that would be a no-op for RW instances)

CraigFe replied:
I think it might be better to raise an exception in the RW case; it's probably indicative of a misunderstanding of the API. Perhaps an argument in favour of phantom capabilities (or a completely different functor instantiation).

icristescu replied:
It now raises an exception, as in index.

samoht commented on src/irmin-pack/pack.ml:
```diff
@@ -276,8 +278,10 @@ struct
+          try
+            let v = unsafe_find t k in
+            Lwt.return v
+          with exn -> Lwt.fail exn)
```
Why do we need an extra `try .. with` here?

icristescu replied:
we don't :)

samoht commented on src/irmin-pack/irmin_pack.ml:
```diff
@@ -448,6 +448,11 @@ struct
+      let ro_sync t =
+        Contents.CA.ro_sync (contents_t t);
+        Node.CA.ro_sync (snd (node_t t));
+        Commit.CA.ro_sync (snd (commit_t t))
```
what happens in the case where the 3 stores are actually the same store? Do we have a way to make this more efficient?

icristescu replied:
I left only one `ro_sync` call and added a comment.

icristescu commented:
If during a merge, an RO instance calls ro_sync and then looks for a value, that was flushed to disk by the merge, it will lead to an exception raised instead of returning None (see https://github.com/mirage/irmin/pull/1008/files#diff-51f1262f591288eadf8719ee5f293b18R309).   

I added some tests that use the `force_merge` function from index, so I switched to `Private.Make` in pack_index. I think this is fine, we are not allowed to use pack_index from outside irmin-pack, but maybe there is a cleaner way to do this?

CraigFe commented on src/irmin-pack/irmin_pack.mli:
```diff
@@ -98,6 +101,9 @@ module Make
       reporting. [`Fixed] and [`Corrupted] report the number of fixed/corrupted
       entries. *)
+
+  val ro_sync : repo -> unit
```
Perhaps a good time to rip out an `EXT` / `Ext` / `Store_extras` module type containing these two functions and include them here with destructive type substitution. We're already duplicating ~10 lines.

icristescu replied:
much cleaner, thanks !

CraigFe commented on src/irmin-pack/irmin_pack.ml:
```diff
@@ -448,6 +448,9 @@ struct
         Node.CA.close (snd (node_t t)) >>= fun () ->
         Commit.CA.close (snd (commit_t t)) >>= fun () -> Branch.close t.branch
+
+      (** stores share on-disk files, one ro_sync is enough *)
```
It's not clear to me why this is true. Don't these store instances contain separate mutable state that needs to be updated? (albeit to the same values in each case.) For instance, will the `Node` instance detect a change to the generation number?

icristescu replied:
the three stores share the same instance of both pack (https://github.com/mirage/irmin/blob/master/src/irmin-pack/pack.ml#L145) and dict (https://github.com/mirage/irmin/blob/master/src/irmin-pack/pack_dict.ml#L18), if I understand this cache correctly.

CraigFe replied:
I certainly don't understand it correctly, so I defer to your judgement :slightly_smiling_face:

CraigFe replied:
Thinking about this a bit more, I think this is good but the comment might need changed:

```suggestion
      (** stores share instances in memory, one ro_sync is enough *)
```

i.e. it's not safe because just the files are the same, it's safe because – in addition – the in-memory mutable state that tracks those fails is identical.

CraigFe commented on src/irmin-pack/dict.ml:
```diff
@@ -78,11 +80,14 @@ module Make (IO : IO.S) : S = struct
 
+  let ro_sync t =
+    if IO.readonly t.io then sync_offset t
+    else Fmt.invalid_arg "only an RO instance should call this function"
```
```suggestion
    else invalid_arg "only an RO instance should call this function"
```

CraigFe commented on test/irmin-pack/test_pack.ml:
```diff
@@ -131,6 +133,20 @@ module Dict = struct
+  val with_thread : (unit -> unit) -> unit
+
+  val await : Index.async -> unit
+end = struct
+  let catch_exn = ref []
+
+  let with_thread f = try f () with exn -> catch_exn := exn :: !catch_exn
+
+  let await thread =
+    Index.await thread;
+    match !catch_exn with [] -> () | e :: _ -> raise e
+end
```
This module confuses me :slightly_smiling_face: `with_thread` doesn't seem to be creating a thread? When we pass this into the log hook, do we not want to re-raise the exception? Is it possible for more than one exception to be accumulated in the list?

icristescu replied:
I copied this module from index without too much thought, and I agree that the names are not great. We can rename `with_thread` to `in_thread`? 
About the exception handling part, I think the motivation was not to lose exceptions generated in the merge thread. Do you think it is not necessary?

CraigFe replied:
> About the exception handling part, I think the motivation was not to lose exceptions generated in the merge thread. Do you think it is not necessary?

It seems like a good idea to fail if a merge thread raises an exception. However, I think this module changes the semantics of the merge thread by causing it _not_ to fail if an exception is thrown: I think we probably want to re-raise the exception after logging it here, so that the thread still fails.

If so, it's (probably ?) the case that we can only ever get at most one exception in the `catch_exn` ref, so it could be a `exception option ref` instead (with an `assert false` if `Some` is detected on update). I think we can't rely on this with multiple asynchronous threads anyway, since the `cons` isn't atomic.

> We can rename `with_thread` to `in_thread`?

Maybe `with_logged_exceptions`? I don't think it has much to do with threads.

icristescu replied:
Thanks, renamed and changed to a catching a single exception.

CraigFe commented on src/irmin-pack/pack_dict.ml:
```diff
@@ -20,4 +20,6 @@ let (`Staged v) =
 let v ?fresh ?readonly ?(capacity = 100_000) root =
-  v capacity ?fresh ?readonly root
+  let dict = v capacity ?fresh ?readonly root in
+  if readonly = Some true then ro_sync dict;
```
What's the purpose of syncing immediately after creating the instance? Is this to do with shared instances? (if so, does that present problems for the "RO instances should not have data changed under their feet" goal?)

icristescu replied:
it's not needed, indeed (removed it).

CraigFe commented on test/irmin-pack/concurrent.ml:
```diff
@@ -0,0 +1,191 @@
+    let m = Bytes.of_string (string_of_int (Unix.getpid ())) in
+    let n = Unix.write input m 0 (Bytes.length m) in
+    assert (n = Bytes.length m)
+  with Unix.Unix_error (n, f, arg) ->
+    Alcotest.failf "Unix error %s %s %s " f arg (Unix.error_message n)
```
Does Alcotest not print such exceptions by default?

icristescu replied:
it does :)

CraigFe commented on test/irmin-pack/concurrent.ml:
```diff
@@ -0,0 +1,191 @@
+
+let worker w_pipe r_pipe =
+  Log.debug (fun l -> l "Worker started %d" (Unix.getpid ()));
+  ignore (read r_pipe);
```
What's being read and ignored here?

icristescu replied:
[The parent pid](https://github.com/mirage/irmin/pull/1008/files#diff-4fdf416b755e60ec1edde6f08f834590R115). Changed it to be more specific.

icristescu commented:
> I'm very uneasy about that. The private API of Index is now being exposed by `Irmin_pack.Index`, and the tests get access to that API very easily (no references to `Private` or such). Indeed, constructing a `Pack` instance now requires the user to construct an `Index` with the private API.
> 
> If testing Index dependencies requires them having access to the private API, perhaps we need to rethink the public API 🙂

I would like to have a `Make_private` functor for irmin pack that can use the private API of index. I tried and it leads to a lot of code duplication, so I gave up. 

But we need to use the private API of index in this PR, the problematic behaviour of `ro_sync` only occurs during a merge.

icristescu commented:
> I added some tests that use the `force_merge` function from index, so I switched to `Private.Make` in pack_index. 

I removed the private API of index, and used `Index.filter` and `Index.flush` instead.

CraigFe commented on src/irmin-pack/dict.mli:
```diff
@@ -23,6 +23,8 @@ module type S = sig
 
   val sync : t -> unit
 
+  val ro_sync : t -> unit
```
We now have both `sync` and `ro_sync` here, doing different things. Perhaps time for a doc-comment / renaming one of them.

icristescu replied:
I renamed `sync` to `flush` (what is called in index); and `ro_sync` to `sync`. Maybe it's better to have `rw_flush` and `ro_sync`? Once we decide on the names, I can change them in index too.

CraigFe commented on test/irmin-pack/concurrent.ml:
```diff
@@ -0,0 +1,187 @@
+
+let run_test () = if Unix.getpid () = parent_pid then Lwt_main.run (check_os ())
+
+(*let () =
+  report ();
+  if Unix.getpid () = parent_pid then Lwt_main.run (check_os ())
+*)
```
Stray comment here.

CraigFe commented on test/irmin-pack/concurrent.ml:
```diff
@@ -0,0 +1,187 @@
+  if Unix.getpid () = parent_pid then Lwt_main.run (check_os ())
+*)
+let tests =
+  [ Alcotest.test_case "Test concurrent reads" `Quick (fun () -> run_test ()) ]
```
```suggestion
  [ Alcotest.test_case "Test concurrent reads" `Quick run_test ]
```

CraigFe commented on test/irmin-pack/concurrent.ml:
```diff
@@ -0,0 +1,187 @@
+      S.Repo.close rw
+
+let check_os () =
+  if Sys.os_type = "Win32" then Lwt.return_unit else concurrent_reads ()
```
Perhaps a comment here to say that the reason we can't run this on Windows is the use of `fork`.

CraigFe commented on test/irmin-pack/test_pack.ml:
```diff
@@ -280,6 +284,77 @@ module Pack = struct
+      Pack.unsafe_append w h3 x3;
+      Index.flush t.index;
+      check h2 x2 "find after flush" >>= fun () ->
+      Pack.sync t.pack;
+      Pack.ro_sync r;
```
Once again, I think here the naming of `sync` vs `ro_sync` is misleading / not very descriptive.

CraigFe commented:
As discussed offline, the tricky case of "flush to index before indexed value has been flushed to pack" could be solved by having Index support a pre-flush callback. We would use that callback to flush to Pack.

CraigFe commented:
(And we should be sure to `fsync` that flush inside Pack, so as to keep the ordering correct even in the presence of crashes.)

icristescu commented:
> As discussed offline, the tricky case of "flush to index before indexed value has been flushed to pack" could be solved by having Index support a pre-flush callback. We would use that callback to flush to Pack.

I added the hook from https://github.com/mirage/index/pull/189 in the latest commit, and the tests now work fine, no invalid_read is raised.

CraigFe commented on src/irmin-pack/pack.ml:
```diff
@@ -191,6 +191,17 @@ struct
+    let set_hook_index index t =
+      Index.set_merge_hook index (fun () ->
+          Dict.flush t.pack.dict;
+          IO.flush t.pack.block)
```
Fsync needed here?

icristescu replied:
I'm thinking this is maybe not necessary. We are not using `fsync` for any pack flush, but instead we use the `integrity_check` to repair the pack/index. I think `fsync` is costly, and as we have the `integrity_check` already, we don't need it. What do you think?

CraigFe replied:
As discussed offline, catching it with `integrity_check` is probably the way to go to avoid having costly flushes. There's now a comment here to explain this intent :+1:

samoht commented on src/irmin-pack/pack.ml:
```diff
@@ -191,6 +191,12 @@ struct
         true )
       else false
 
+    let flush ?(all = true) t =
```
maybe `index` instead of `all` to make clear that this flag only control the flush of the underlying index.

CraigFe commented on CHANGES.md:
```diff
@@ -1,3 +1,9 @@
+### Unreleased
+
+#### Changed
+- `sync` has to be called by the read-only instance to synchronise with the
+  files on disk. (#1008)
```
```suggestion
- **irmin-pack**
  - `sync` has to be called by the read-only instance to synchronise with the
     files on disk. (#1008)
```

This PR also changes the meaning of `sync` (which is now `flush`). Perhaps need another sentence to explain that.

icristescu replied:
Thanks, added a sentence about flush.

CraigFe commented on test/irmin-pack/test_pack.ml:
```diff
@@ -280,6 +284,65 @@ module Pack = struct
         Alcotest.fail "Add after closing the index should not be allowed")
       (function I.Closed -> Lwt.return_unit | exn -> Lwt.fail exn)
 
+  (** Index can be flushed to disk independently of pack, we simulate this in
+      the tests using [Index.filter] and [Index.flush]. *)
```
Can you add some explanation about why this might be an issue? E.g. "Regression test for PR 1008 in which values were indexed before being reachable in Pack."

icristescu replied:
Added your suggested comment.

CraigFe commented:
Thanks. Will wait for Travis to wake up and then merge.
